lib/mix/lib/mix/compilers/elixir.ex

```diff
@@ -32,34 +32,37 @@ defmodule Mix.Compilers.Elixir do
     # change to files are still picked up by the compiler. This
     # timestamp is used when writing BEAM files and the manifest.
     timestamp = :calendar.universal_time()
-    all = Mix.Utils.extract_files(srcs, [:ex])
+    all_paths = MapSet.new(Mix.Utils.extract_files(srcs, [:ex]))
 
     {all_modules, all_sources} = parse_manifest(manifest, dest)
     modified = Mix.Utils.last_modified(manifest)
+    prev_sources_paths =
+      for source(source: source) <- all_sources, into: MapSet.new(), do: source
 
     removed =
-      for source(source: source) <- all_sources,
-          not(source in all),
-          do: source
+      prev_sources_paths
+      |> MapSet.difference(all_paths)
+      |> MapSet.to_list
 
     changed =
       if force do
         # A config, path dependency or manifest has
         # changed, let's just compile everything
-        all
+        MapSet.to_list(all_paths)
       else
         sources_mtimes = mtimes(all_sources)
 
         # Otherwise let's start with the new sources
-        for(source <- all,
-            not List.keymember?(all_sources, source, source(:source)),
-            do: source)
-          ++
+        new_paths =
+          all_paths
+          |> MapSet.difference(prev_sources_paths)
+          |> MapSet.to_list
         # Plus the sources that have changed in disk
-        for(source(source: source, external: external) <- all_sources,
-            times = Enum.map([source | external], &Map.fetch!(sources_mtimes, &1)),
-            Mix.Utils.stale?(times, [modified]),
-            do: source)
+        changed_mtime = for(source(source: source, external: external) <- all_sources,
```

Comment:
We usually would format this as follows:
```elixir
changed_mtime =
  for(source(source: source, external: external) <- all_sources,
      times = Enum.map([source | external], &Map.fetch!(sources_mtimes, &1)),
      Mix.Utils.stale?(times, [modified]),
      do: source)
```
We can actually get rid of `++` with:
```elixir
for(source(source: source, external: external) <- all_sources,
    times = Enum.map([source | external], &Map.fetch!(sources_mtimes, &1)),
    Mix.Utils.stale?(times, [modified]),
    into: new_paths,
    do: source)
```
_Note `:into` option._